### PR TITLE
Refresh features in Attribute Table on data change

### PR DESCRIPTION
**Pull request by signedav**
On adding / removing / changing features.

- On filter "Show Visible" the list is updated
- On custom filters the list is updated

(since the update needs to be done on data changed and not only on filter changed, I moved the filter functionality to the model instead the widget)

- update of count of features in the AttributeEdit mode of the DualView, on FeatureListModel adding or removing feature

### CODE REVIEW

File: src/gui/attributetable/qgsattributetablefiltermodel.cpp [hunk 1]
```diff
@@ -371,12 +383,64 @@ bool QgsAttributeTableFilterModel::filterAcceptsRow( int sourceRow, const QModel
...
   invalidateFilter();
 }
 
+void QgsAttributeTableFilterModel::filterFeatures()
+{
+  if ( !mFilterExpression.isValid() )
+    return;
+
+  QgsFeatureIds filteredFeatures;
+  QgsDistanceArea myDa;
```
Comment (elpaso) [inline src/gui/attributetable/qgsattributetablefiltermodel.cpp]:
```suggestion
  QgsDistanceArea distanceArea;
```

File: src/gui/attributetable/qgsattributetablefiltermodel.cpp [hunk 2]
```diff
@@ -371,12 +383,64 @@ bool QgsAttributeTableFilterModel::filterAcceptsRow( int sourceRow, const QModel
...
+
+  myDa.setSourceCrs( mTableModel->layer()->crs(), QgsProject::instance()->transformContext() );
+  myDa.setEllipsoid( QgsProject::instance()->ellipsoid() );
+
+  bool fetchGeom = mFilterExpression.needsGeometry();
```
Comment (elpaso) [inline src/gui/attributetable/qgsattributetablefiltermodel.cpp]:
```suggestion
  const bool fetchGeom = mFilterExpression.needsGeometry();
```

File: src/gui/attributetable/qgsattributetablefiltermodel.h [hunk 3]
```diff
@@ -221,6 +221,12 @@ class GUI_EXPORT QgsAttributeTableFilterModel: public QSortFilterProxyModel, pub
      */
     void setAttributeTableConfig( const QgsAttributeTableConfig &config );
 
+    /**
+     * Set the \a expression to be stored in case of the features need to be
+     * filtered (like on filter or on main model data change).
+     */
+    void setFilterExpression( const QgsExpression &expression, const QgsExpressionContext &context ) { mFilterExpression = expression; mFilterExpressionContext = context; }
```
Comment (elpaso) [inline src/gui/attributetable/qgsattributetablefiltermodel.h]: move to cpp file and add missing docs for context

File: src/gui/attributetable/qgsattributetablefiltermodel.h [hunk 4]
```diff
@@ -255,10 +261,13 @@ class GUI_EXPORT QgsAttributeTableFilterModel: public QSortFilterProxyModel, pub
...
     /**
-     * Is called upon every change of the visible extents on the map canvas.
+     * Is called upon every change of the visible extents on the map canvas or when data of
+     * the master table model changes.
      * When a change is signalled, the filter is updated and invalidated if needed.
      */
-    void extentsChanged();
+    void reloadVisible();
+
+    void filterFeatures();
```
Comment (elpaso) [inline src/gui/attributetable/qgsattributetablefiltermodel.h]: missing docs

File: src/gui/attributetable/qgsattributetablefiltermodel.h [hunk 5]
```diff
@@ -255,10 +261,13 @@ class GUI_EXPORT QgsAttributeTableFilterModel: public QSortFilterProxyModel, pub
   public slots:
 
     /**
-     * Is called upon every change of the visible extents on the map canvas.
+     * Is called upon every change of the visible extents on the map canvas or when data of
+     * the master table model changes.
```
Comment (elpaso) [inline src/gui/attributetable/qgsattributetablefiltermodel.h]:
```suggestion
     * the master table model change.
```

File: src/gui/attributetable/qgsfeaturelistview.h [hunk 6]
```diff
@@ -214,6 +214,11 @@ class GUI_EXPORT QgsFeatureListView : public QListView
   private slots:
     void editSelectionChanged( const QItemSelection &deselected, const QItemSelection &selected );
 
+    /**
+     * Emmits the signal for the feature and the selection information
```
Comment (elpaso) [inline src/gui/attributetable/qgsfeaturelistview.h]:
```suggestion
     * Emits the signal for the feature and the selection information
```

**Review by elpaso — CHANGES_REQUESTED** [review]

**Comment by signedav** [comment]
Thanks @elpaso I added a small test as well...

**Comment by signedav** [comment]
What do you think about a backport @elpaso It's imho a bugfix, but not critical...

**Comment by signedav** [comment]
@elpaso is this good to merge?

**Pull request merged**